### DISCUSSION TITLE
Enfants d'une règle `non applicable`

### DISCUSSION THREAD

**Discussion by Clemog** [publicodes/publicodes#373] [V2]
Hello ! Je ne sais pas très bien s'il fallait faire une issue ou discussions.

Je voulais remonter un problème déjà remonté à plusieurs reprises mais de mon côté c'est très problématique:

Aujourd'hui, le mécanisme `non applicable` est designé de sorte que les enfants d'une règles `non applicable` devient non applicable également. Dans Nos Gestes Climat, on utilise énormément ce mécanisme et nous utilisons également beaucoup de règles dépendantes d'un mécanisme `non applicable` "ailleurs". 

Comme ici : https://publi.codes/studio/d#a%3A%0A%20%20applicable%20si%3A%20b%0A%20%20formule%3A%2010%0A%0Ab%3A%20non%0A%0Aa%20.%20c%3A%2010%0A%0Ad%3A%20a%20.%20c

Pour contourner ce pb, j'utilise en général des `variations` mais ce n'est pas idéal ..

https://publi.codes/studio/d#a%3A%0A%20%20formule%3A%0A%20%20%20%20variations%3A%0A%20%20%20%20%20%20-%20si%3A%20b%0A%20%20%20%20%20%20%20%20alors%3A%2010%0A%20%20%20%20%20%20-%20sinon%3A%200%0A%0Ab%3A%20non%0A%0Aa%20.%20c%3A%2010%0A%0Ad%3A%20a%20.%20c

Il me semble que dans l'historique, ce comportement a été implémenté assez "tard" et je me rends compte parfois d'erreur liées à ça dans nos actions.

Je ne retrouve pas les discussions sur le sujet, peut etre que ça a déjà été longuement débattu alors désolé pour le doublon.. Sinon je veux bien en discuter davantage pour voir comment améliorer ça ;)

ping @laem @johangirod

**Comment by johangirod** [comment]
C'est un comportement implémenté depuis très longtemps, à mon arrivée en 2018. L'idée originale était de simplifier la base de code qui contenait beaucoup de `applicable si: parent`. Cela nous semblait un ajout simple et facile pour éviter de se répéter. 

Mais, quelques années plus tard, je te rejoins sur le fait que ce comportement peut prêter à confusion. Par ailleurs, il nous a souvent gênés, @mquandalle et moi, lors de précédentes refacto.

Dans l'idée, je préfèrerais que le parent rende explicitement non applicable ses enfants. Ce serait un comportement plus clair, et moins surprenant. Pour ce faire, on pourrait imaginer étendre la syntaxe de `rend non applicable` avec des sélecteurs qui vise tous les enfants. Par exemple : 

```yaml
a:
  applicable si: b
  rend non applicable: a . *
```
ou encore

```yaml
a:
  applicable si: b
  rend non applicable: 
    toutes les règles de: a
```

Cela aurait l'avantage de résoudre le problème listé dans https://github.com/betagouv/publicodes/issues/206

Il y a néanmoins deux problèmes avec ce changement : 
1. C'est un gros changement cassant, qui suppose beaucoup de modifications assez fine des règles publicodes existantes pour ne pas engendrer de régressions. 
2. `rend non applicable` fonctionne comme le remplacement. Or les remplacements transitifs ne sont pas implémentés. Cela aboutirait à un bug dans le cas d'une règle rendue non applicable et remplacée. Il faudrait donc corriger #55 avant. 

En attendant, la solution intermédiaire est de déplacer les valeurs dans un autre espace de nom. C'est ce que l'on fait sur mon-entreprise.

**Comment by Clemog** [comment]
> C'est un comportement implémenté depuis très longtemps, à mon arrivée en 2018. L'idée originale était de simplifier la base de code qui contenait beaucoup de `applicable si: parent`.

Ah d'acc, je suis étonné car j'ai l'impression que le changement du comportement du `applicable si` était plus récent donc sorry :)

Merci pour #206 !

Je me demande juste pour quelle raison il a été décidé que toute la branche soit non applicable si le parent de ne l'est pas au départ ?

Je ne suis pas sur de bien comprendre tes 2 points

> En attendant, la solution intermédiaire est de déplacer les valeurs dans un autre espace de nom. C'est ce que l'on fait sur mon-entreprise.

Yes également !

**Comment by laem** [comment]
@Clemog est-ce que tu as un exemple plus complet, réel pour décrire le problème ?

Qu'est-ce que `c` ? Un facteur d'émission que tu as rangé dans l'espace de nom `voiture` mais que tu voudrais utiliser ailleurs ? 

Si c'est le cas, alors la solution la plus simple à mon avis serait de permettre un `fetch`. Quelle que soit la valeur de l'applicabilité évaluée, tu veux juste récupérer la valeur de l'enfant. 

En termes d'historique, je crois que ça a été décidé notamment pour alléger fortement le calcul : aucune raison de calculer les enfants d'un non applicable, c'est du temps perdu. 

Un `fetch` pourrait justement vouloir dire : `Certes toute la branche peut être non applicable, mais là j'ai besoin que tu évalues quand même juste cet enfant, indépendamment de l'applicabilité de ses parents`.

**Comment by laem** [comment]
Après, je suis plus au courant de l'implé actuelle, mais dans ton exemple initial, il semble que de toutes façon `c` est calculée. En tout cas son calcul s'affiche dans la doc. Donc il y aurait juste un blocage fait au niveau de l'appel de `a . c` qui retourne `non applicable` car `a` l'est ? Dans ce cas là l'appel via un mécanisme ferait l'affaire sans trop de changements. 

```yaml
a:
  applicable si: b
  formule: 10

b: non

a . c: 23 + 6

d: 
  valeur isolée: a . c
```

**Comment by Clemog** [comment]
> @Clemog est-ce que tu as un exemple plus complet, réel pour décrire le problème ?

J'ai mis un exemple plus haut mais plus concrètement : la mosaïque sur les éco gestes contient `gestes . bonus stop pub`. gestes est `non applicable` sous certaines conditions donc `bonus stop pub` n'est pas calculé alors que c'est la valeur d'une action `applicable` justement quand la mosaïque ne l'est pas

(https://github.com/datagir/nosgestesclimat/blob/2c419c5b239d1f83c9417be443768b63554b4bc7/data/alimentation/mod%C3%A8le%20d%C3%A9chets/alimentation%20.%20d%C3%A9chets.yaml#L185)

**Comment by johangirod** [comment]
> Un fetch pourrait justement vouloir dire : Certes toute la branche peut être non applicable, mais là j'ai besoin que tu évalues quand même juste cet enfant, indépendamment de l'applicabilité de ses parents.

C'est techniquement déjà possible avec le mécanisme `recalcul`, même si la syntaxe est alors un peu lourde. 

> Ah d'acc, je suis étonné car j'ai l'impression que le changement du comportement du applicable si était plus récent donc sorry :)

On a revu l'implémentation de l'applicabilité, mais ce comportement était présent. 

> Je ne suis pas sur de bien comprendre tes 2 points

On peut en rediscuter de vive voix, si tu le souhaites.

**Comment by Clemog** [comment]
Je réouvre un peu le sujet ;)

> Après, je suis plus au courant de l'implé actuelle, mais dans ton exemple initial, il semble que de toutes façon `c` est calculée. En tout cas son calcul s'affiche dans la doc. Donc il y aurait juste un blocage fait au niveau de l'appel de `a . c` qui retourne `non applicable` car `a` l'est ? Dans ce cas là l'appel via un mécanisme ferait l'affaire sans trop de changements.

Hum, il ne me semble pas que ce soit le cas, à priori tu étais dans un cas particulier de la somme de 2 constante directement dans la formule.

Exemple [ici](https://publi.codes/studio/a/c#a%3A%0A%20%20applicable%20si%3A%20b%0A%20%20formule%3A%2010%0A%0Ab%3A%20non%0A%0Aa%20.%20c%3A%2023%20%2B%20d%0A%0Aa%20.%20d%3A%0A%20par%20défaut%3A%209%0A%0Ae%3A%20%0A%20%20valeur%20isolée%3A%20a%20.%20c) (En supposant par exemple que `d` est une question.)

**Comment by Clemog** [comment]
Je me dis aussi qu'on pourrait partir sur un fonctionnement à l'inverse de ce que tu proposes @johangirod en gardant le comportement actuel `par défaut` mais qui force l'applicabilité de règles spécifiées ? Implicitement, le mécanisme serait un recalcul dans lequel on rend la variable qui rend la règle non applicable, applicable ou bien il y a un fonctionnement plus simple ?

```yaml
a:
  applicable si: b
  
a . b:
  question: oui ou non ? 
  par défaut: non
  
a . c:
  titre: la règle que l'on veut appeler ailleurs
  formule: d * e

a . c . d: 5

a . c . e: 10

c prime:
  recalcul:
    règle: a . c
    avec:
      a . b: oui
```

(Dans l'exemple [ici](https://publi.codes/studio/c-prime#a%3A%0A%20%20applicable%20si%3A%20b%0A%23%20%20rend%20non%20applicable%3A%20a%20.%20*%0A%20%20%0Aa%20.%20b%3A%0A%20%20question%3A%20oui%20ou%20non%20%3F%20%0A%20%20par%20défaut%3A%20non%0A%20%20%0Aa%20.%20c%3A%0A%20%20titre%3A%20la%20règle%20que%20l'on%20veut%20appeler%20ailleurs%0A%20%20formule%3A%20d%20*%20e%0A%0Aa%20.%20c%20.%20d%3A%205%0A%0Aa%20.%20c%20.%20e%3A%2010%0A%0Ac%20prime%3A%0A%20%20recalcul%3A%0A%20%20%20%20règle%3A%20a%20.%20c%0A%20%20%20%20avec%3A%0A%20%20%20%20%20%20a%20.%20b%3A%20oui), on a bien la valeur de c dans c prime)

Je propose par exemple :

```yaml
a:
  applicable si: b
  restent applicable:
    - a . c . *
    - a . d
```

Dans ce cas on aurait :

- Toutes les règles de `c`(`c` compris) deviendraient applicables
- Seule d (et ses dépendances de calcul) deviendraient applicable
- Si `a` possède une règle `e`, alors `a . e` reste non applicable


Un point d'attention (je crée peut être des faux pb):

- Attention à ne pas rendre applicable artificiellement des règles qui ne devraient vraiment pas l'être : ex si `c` est applicable selon certaines conditions


> * C'est un gros changement cassant, qui suppose beaucoup de modifications assez fine des règles publicodes existantes pour ne pas engendrer de régressions.
> * `rend non applicable` fonctionne comme le remplacement. Or les remplacements transitifs ne sont pas implémentés. Cela aboutirait à un bug dans le cas d'une règle rendue non applicable et remplacée. Il faudrait donc corriger [Transitivité du remplacement #55](https://github.com/betagouv/publicodes/issues/55) avant.

- J'ai l'impression qu'avec ce type de fonctionnement, on résout ton premier point ?
- Je regarde pour les remplacements transitifs mais je ne suis pas bien sur de comprendre

**Reply by johangirod** [reply]
Oui, ça me parait bien ! Le mot clé est clair, est ne devrait pas être compliqué à implémenter.

**Comment by Clemog** [comment]
Et pour [ce cas précis](https://publi.codes/studio/titre%E2%80%91restau#titre-restau%3A%0A%20%20applicable%20si%3A%0A%20%20%20%20nom%3A%20éligible%0A%20%20%20%20question%3A%20Le%20salarié%20reçoit-il%20des%20titres-restaurant%20%3F%0A%20%20%20%20par%20défaut%3A%20non%0A%20%20formule%3A%204€) de #206, il me semble que nous n'avons plus ce pb ?

**Comment by Clemog** [comment]
Petit up sur le sujet, j'avais un peu laissé de côté cette issue qui est toujours d'actu pour nous, notamment pour l'affichage de questions booléennes qui sont des règles de la condition d'applicabilité du parent :

```
a: 
  applicable si: a . b
  
a . b: 
  question: à poser
  valeur: non
```

Dans ce cas, b est non applicable et on est "coincés", une solution facile serait de décaler l'espace nom mais ça m'embête ..

On etait plutôt d'accord pour cette solution : https://github.com/publicodes/publicodes/discussions/373#discussioncomment-6595119

J'en reparle jeudi au café :)

(@mquandalle si tu as le temps de jeter un oeil avant, je veux bien ton avis :) )

**Reply by mquandalle** [reply]
Je n'ai pas rencontré ce problème ni sur mesaidesvelo ni sur mon-entreprise. J'ai toujours organisé les règles et questions en prenant en compte qu'une question dans un espace de nom désactivable serait cachée si l'espace de nom est désactivé. C'est en général le comportement que je veux, par exemple si j'ai une question `aide . dijon . vélo fabriqué localement` pour affiner le calcul d'une aide vélo à Dijon, mais que la règle `aide . dijon` n'est pas active je veux bien ignorer la question. Et si je veux garder une question toujours active je ne l'a met pas dans un espace de nom désactivable.

Tu as un fichier complet réel pour illustrer l'intérêt de déplacer des questions toujours actives dans des espaces de noms désactivables ?

**Reply by Clemog** [reply]
Oui je comprends
Par habitude (je pense) et par desgin au départ, on a toujours laissé nos règles conditionnelles dans l'espace nom qu'elle rend applicable ou non (exemple avec les [deux roues](https://github.com/incubateur-ademe/nosgestesclimat/blob/preprod/data/transport/deux%20roues.publicodes) même si c'est un tout petit peu plus complexe car on a une variation porur la valeur par défaut).

**Reply by Clemog** [reply]
On a également fait le travail inverse en faisant en sorte de sortir de l'espace nom la variable qu'on rend applicable (ou non), pour être sur d'avoir accès à une règle à tout moment en étant sur qu'elle ne sera jamais désactivée (genre le le facteur d'émission du gaz [ici](https://github.com/incubateur-ademe/nosgestesclimat/blob/5cf5f5ec41b841894ee272a35e0ee3750979e5e3/data/logement/chauffage.publicodes#L95))

**Reply by Clemog** [reply]
Avec l'attribut `reste applicable`, on éviterait la gymnastique en désctivant "gaz" si non applicable **sauf** les règle qu'on souhaite garder (genre le facteur d'émission)

**Reply by Clemog** [reply]
(Je viens de tester en sortant la règle du parent non applicable pour les 2 roues par exemple mais ça casse l'ordre de nos questions basé sur les espaces noms.. donc j'ai l'"impression que seul le mécanisme pourrait aider)

**Comment by mquandalle** [comment]
On a rediscuté de ce point avec @Clemog aujourd'hui.
Sur la désactivation de branche, je n'ai personnellement pas rencontré ce problème lié à l'organisation des questions, mais il semble raisonnable d'utiliser la fonctionnalité d'espace de noms pour profiter de références plus concises, sans que cela n'impacte tacitement l'applicabilité des règles de l'espace. Nous avons vu que ça pouvait être une source de bug pour des contributeurs non familiers avec ce fonctionnement.

Même si la fonctionnalité est ergonomique à l'usage, l'implémentation de la désactivation de branche n'est vraiment pas simple, en particulier du fait qu'il faut gérer la composition avec les autres fonctionnalités de publicodes comme les remplacements. Je suis donc plutôt défavorable à la complexifier encore avec un attribut type "ne pas désactiver les règles enfants suivantes" ou au niveau de la règle enfant un attribut "règle toujours applicable même si le parent est désactivé".

En revanche ça me semble de bon principe d'expliciter la désactivation de branche au niveau où il s'applique. Quelque chose comme:

```yaml
a:
  espace applicable si: condition
  valeur: 3
```

Ça rend plus explicite que la condition impacte tout l'espace de nom et non la seule règle. Je soupçonne que ça soit aussi positif pour les performances, un certains temps étant actuellement passé à insérer des références virtuelles aux parents et à les évaluer au cas où ils seraient non applicable — ce qui veut dire qu'un utilisateur qui n'utilise pas la désactivation de branche en paye quand même le prix en temps de calcul.

Pas sûr de la manière la plus élégante d'expliciter cette désactivation de branche, ma proposition impliquerait de décliner tous les mécanismes d'applicabilité dans une version "espace": `espace applicable si`, `espace non applicable si`, `rend l'espace non applicable`...

**Reply by mquandalle** [reply]
Pour l'implémentation si on pouvait supprimer cette condition ça serait bien (empêcher les boucles infinies lors de l'évaluation en vérifiant si la règle appelée n'est pas déjà dans la pile)

https://github.com/publicodes/publicodes/blob/master/packages/core/src/rule.ts#L206-L209


Au lieu de ça on devrait avoir une garantie statique qu'il n'y a pas de cycle du fait de la désactivation d'espaces.

**Reply by mquandalle** [reply]
Et #206 devrait probablement être traité dans le même temps.

**Reply by johangirod** [reply]
J'aime beaucoup cette solution. C'est un gros changement cassant, mais je pense que l'on peut l'automatiser en partie avec des codemod un peu intelligents. À prévoir dans une version 2, donc ?

**Comment by Clemog** [comment]
Merci pour le retour @mquandalle :)

Je comprends ton point, il me semble qu'effectivement, ce n'est pas contre intuitif de partir du fait que "l'espace" complet devient non applicable si la parent l'est. 

On passerait donc sur [ce type d'écriture](https://publi.codes/studio/a#a%3A%0A%20%20somme%3A%0A%20%20%20%20-%20variable%20constante%0A%20%20%20%20-%20d%0A%0Aa%20.%20b%3A%20%0A%20%20question%3A%20coucou%0A%20%20valeur%3A%20non%0A%0Aa%20.%20c%3A%20%0A%20%20titre%3A%20on%20en%20a%20besoin%20ailleurs%0A%20%20valeur%3A%2010%0A%0Aa%20.%20d%3A%0A%20%20applicable%20si%3A%20b%0A%20%20question%3A%20coucou%0A%0Aa%20.%20variable%20constante%3A%0A%20%20applicable%20si%3A%20b%0A%20%20formule%3A%2010) comme on se disait hier.

Néanmoins, je propose [un autre exemple](https://publi.codes/studio/a#a%3A%0A%20%20formule%3A%0A%20%20%20%20variations%3A%0A%20%20%20%20%20%20-%20si%3A%20%0A%20%20%20%20%20%20%20%20%20%20est%20d%C3%A9fini%3A%20b%0A%20%20%20%20%20%20%20%20alors%3A%20ma%20somme%0A%20%20%20%20%20%20-%20sinon%3A%20autre%20valeur%0A%0Aa%20.%20autre%20valeur%3A%0A%20%20formule%3A%2010%0A%0Aa%20.%20ma%20somme%3A%0A%20%20applicable%20si%3A%20b%0A%20%20formule%3A%0A%20%20%20%20somme%3A%0A%20%20%20%20%20%20-%20variable%20constante%0A%20%20%20%20%20%20-%20d%0A%0Aa%20.%20b%3A%20%0A%20%20question%3A%20coucou%0A%20%20valeur%3A%20non%0A%0Aa%20.%20c%3A%20%0A%20%20titre%3A%20on%20en%20a%20besoin%20ailleurs%0A%20%20valeur%3A%2010%0A%0Aa%20.%20d%3A%0A%20%20question%3A%20coucou%0A%20%20formule%3A%20100%0A%0Aa%20.%20variable%20constante%3A%0A%20%20formule%3A%2010) (j'ai du mal à voir si c'est incompatible avec ta proposition). On réfléchissait avec @EmileRolley et on voit ici que l'espace nom et la variable sont confondues. `a` devient non applicable (et donc l'espace complet) alors qu'on souhaite seulement attribuer "la valeur" de `a . ma somme` à `a`. Est ce que c'est "par design" ?

**Reply by Clemog** [reply]
Exemple beaucoup plus simple pour le dernier point : [studio](https://publi.codes/studio/a#a%3A%0A%20%20formule%3A%20a%20.%20b%0A%0Aa%20.%20b%3A%0A%20%20applicable%20si%3A%20non%0A%20%20formule%3A%2010%0A%0Aa%20.%20c%3A%0A%20%20formule%3A%201000)

**Reply by mquandalle** [reply]
Si on a `a: applicable si: non` et `b: a` alors il est bien attendu que `b` prenne la valeur et l'applicabilité de `a` [studio](https://publi.codes/studio/b#a%3A%0A%20%20applicable%20si%3A%20non%0A%0Ab%3A%20a).

Le mécanisme somme "absorbe" les termes non applicables, c'est à dire qu'une somme est toujours applicable, et que les termes non applicables sont comptés pour `0` [studio](https://publi.codes/studio/b#a%3A%0A%20%20applicable%20si%3A%20non%0A%0Ab%3A%20%0A%20%20somme%3A%0A%20%20%20%20-%20a%0A). Cela peut permettre de référencer une règle non applicable et de la compter pour la valeur numérique "0".

En fait une règle non applicable se comporte comme le type `null` de JavaScript : 


```js
a = null
b = a // null
c = a + null // 0
```

**Reply by Clemog** [reply]
Ok je vois :)
Donc ce genre de [hack](https://publi.codes/studio/a#a%3A%0A%20%20formule%3A%200%20%2B%20a%20.%20b%0A%0Aa%20.%20b%3A%0A%20%20applicable%20si%3A%20non%0A%20%20formule%3A%2010%0A%0Aa%20.%20c%3A%0A%20%20formule%3A%201000) fonctionnerait 😅 (ou celui que tu partages désolé, j'ai lu trop vite)
